lib/Dialect/RTG/Transforms/LinearScanRegisterAllocationPass.cpp

```diff
@@ -0,0 +1,169 @@
+//===----------------------------------------------------------------------===//
+//
+// This pass allocates registers using a simple linear scan algorithm.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/RTG/IR/RTGISAAssemblyOpInterfaces.h"
+#include "circt/Dialect/RTG/IR/RTGOps.h"
+#include "circt/Dialect/RTG/Transforms/RTGPasses.h"
+#include "mlir/IR/PatternMatch.h"
+#include "llvm/Support/Debug.h"
+
+namespace circt {
+namespace rtg {
+#define GEN_PASS_DEF_LINEARSCANREGISTERALLOCATIONPASS
+#include "circt/Dialect/RTG/Transforms/RTGPasses.h.inc"
+} // namespace rtg
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+
+#define DEBUG_TYPE "rtg-linear-scan-register-allocation"
+
+namespace {
+
+/// Represents a register and its live range.
+struct RegisterLiveRange {
+  rtg::RegisterAttrInterface fixedReg;
+  rtg::VirtualRegisterOp regOp;
+  unsigned start;
+  unsigned end;
+};
+
+class LinearScanRegisterAllocationPass
+    : public circt::rtg::impl::LinearScanRegisterAllocationPassBase<
+          LinearScanRegisterAllocationPass> {
+public:
+  void runOnOperation() override;
+};
+
+} // end namespace
+
+static void expireOldInterval(SmallVector<RegisterLiveRange *> &active,
+                              RegisterLiveRange *reg) {
+  // TODO: use a better datastructure for 'active'
+  llvm::sort(active, [](auto *a, auto *b) { return a->end < b->end; });
+
+  for (auto *iter = active.begin(); iter != active.end(); ++iter) {
+    auto *a = *iter;
+    if (a->end >= reg->start)
+      return;
+
+    active.erase(iter--);
+  }
+}
+
+void LinearScanRegisterAllocationPass::runOnOperation() {
+  auto testOp = getOperation();
+
+  LLVM_DEBUG(llvm::dbgs() << "=== Processing test @" << testOp.getSymName()
+                          << "\n\n");
+
+  DenseMap<Operation *, unsigned> opIndices;
+  unsigned maxIdx;
+  for (auto [i, op] : llvm::enumerate(*testOp.getBody())) {
+    // TODO: ideally check that the IR is already fully elaborated
+    opIndices[&op] = i;
+    maxIdx = i;
+  }
+
+  // Collect all the register intervals we have to consider.
+  SmallVector<std::unique_ptr<RegisterLiveRange>> regRanges;
+  SmallVector<RegisterLiveRange *> active;
+  for (auto &op : *testOp.getBody()) {
+    if (!isa<rtg::FixedRegisterOp, rtg::VirtualRegisterOp>(&op))
+      continue;
+
+    RegisterLiveRange lr;
+    lr.start = maxIdx;
+    lr.end = 0;
+
+    if (auto regOp = dyn_cast<rtg::VirtualRegisterOp>(&op))
+      lr.regOp = regOp;
+
+    if (auto regOp = dyn_cast<rtg::FixedRegisterOp>(&op))
+      lr.fixedReg = regOp.getReg();
+
+    for (auto *user : op.getUsers()) {
+      if (!isa<rtg::InstructionOpInterface>(user)) {
+        user->emitError("only operations implementing 'InstructionOpInterface "
+                        "are allowed to use registers");
+        return signalPassFailure();
+      }
+
+      // TODO: support labels and control-flow loops (jumps in general)
```

Comment:
yea, this is interesting.  We need to be able to deal with fixed-value jumps, jumptables, label relative jumps, etc.  We will probably need an op interface for this.  Something to resolve computed live ranges.